src/AsyncEventSource.cpp

```diff
@@ -190,29 +185,61 @@ AsyncEventSourceClient::~AsyncEventSourceClient() {
 }
 
 bool AsyncEventSourceClient::_queueMessage(const char* message, size_t len) {
-  if (!_client)
+  if (_messageQueue.size() >= SSE_MAX_QUEUED_MESSAGES) {
+#ifdef ESP8266
+    ets_printf(String(F("ERROR: Too many messages queued\n")).c_str());
+#elif defined(ESP32)
+    log_e( "Event message queue overflow: discard message" );
+#endif
     return false;
+  }
 
 #ifdef ESP32
   // length() is not thread-safe, thus acquiring the lock before this call..
   std::lock_guard<std::mutex> lock(_lockmq);
 #endif
 
+  _messageQueue.emplace_back(message, len);
+
+/*
+  throttle queue run
+  if Q is filled for >25% then network/CPU is congested, since there is no zero-copy mode for socket buff
+  forcing Q run will only eat more heap ram and blow the buffer, let's just keep data in our own queue
+  the queue will be processed at least on each onAck()/onPoll() call from AsyncTCP
+*/
```

Comment:
- `SSE_MAX_QUEUED_MESSAGES` should be increased by default  maybe ?

- Even if I understand the goal of this change, its effectiveness really depends on the network speed and the kinf of messages enqueues. For exemple, if an app sends a lot of small messages on a slow network, the queue would fill quickly to 50%, but then when pcb space becomes available, all the queue could be sent at once because of the smaller message sizes.

I suspect that for it to work fine, you would need to calculate the whole queue size and compare it to `client.getMss()` (which might be similar to `CONFIG_TCP_MSS` or `CONFIG_LWIP_TCP_MSS`) to check if the whole queue can be contained in next bcp space....

Reply:
I do not have a lot of experience with SSE's actually, and this kind of crafting is really app specific.
From what I understood doing some simple tests with the example I've provided AWS's implementation mostly does _not_ use queue at all! So I doubt that `SSE_MAX_QUEUED_MESSAGES` is even needed that large.
I.e. on the first added SINGLE message the `AsyncEventSourceClient::_runQueue` is called in the context of user's `send` function. So the message is immediately written to the sock buff even before return to user function no matter how small the message is.
It only starts to queue if user code pushes so large messages that the next user call would happen before first ack would return from AsyncTCP.

> its effectiveness really depends on the network speed and the kinf of messages enqueued
exactly! but the main idea of this throttle is actually completely different :)
the thing is - current lwip's configuration does only deep copy to socket buff, so during the time of congestion for bandwidth we would end end up with TWO data copies - one in `AsyncEventSourceMessage` object until it is acked and _another_ in socket buff. So the question is - why to try to fill socket buff with as much data as possible if any way it is congested? It could simply wait in the chain of `AsyncEventSourceMessage` and refill socket buff on acks.

Reply:
ok! I understand more the logic behind.